providers/aws-ec2/src/main/java/org/jclouds/aws/ec2/features/InternetGatewayApi.java

```diff
@@ -0,0 +1,141 @@
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jclouds.aws.ec2.features;
+
+import static org.jclouds.aws.reference.FormParameters.ACTION;
+
+import javax.inject.Named;
+import javax.ws.rs.FormParam;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+
+import org.jclouds.Fallbacks.EmptyFluentIterableOnNotFoundOr404;
+import org.jclouds.Fallbacks.FalseOnNotFoundOr404;
+import org.jclouds.aws.ec2.binders.BindInternetGatewayIdsToIndexedFormParams;
+import org.jclouds.aws.ec2.domain.InternetGateway;
+import org.jclouds.aws.ec2.options.InternetGatewayOptions;
+import org.jclouds.aws.ec2.xml.DescribeInternetGatewaysResponseHandler;
+import org.jclouds.aws.ec2.xml.InternetGatewayHandler;
+import org.jclouds.aws.ec2.xml.ReturnValueHandler;
+import org.jclouds.aws.filters.FormSigner;
+import org.jclouds.javax.annotation.Nullable;
+import org.jclouds.location.functions.RegionToEndpointOrProviderIfNull;
+import org.jclouds.rest.annotations.BinderParam;
+import org.jclouds.rest.annotations.EndpointParam;
+import org.jclouds.rest.annotations.Fallback;
+import org.jclouds.rest.annotations.FormParams;
+import org.jclouds.rest.annotations.RequestFilters;
+import org.jclouds.rest.annotations.VirtualHost;
+import org.jclouds.rest.annotations.XMLResponseParser;
+
+import com.google.common.collect.FluentIterable;
+
+/**
+ * Provides access to InternetGateway Services.
+ *
+ * @see <a href="http://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_InternetGateway.html">InternetGateway docs</a>
+ * <p/>
+ */
+@RequestFilters(FormSigner.class)
+@VirtualHost
+@Path("/")
+public interface InternetGatewayApi {
+
+   /**
+    * Detaches an {@link InternetGateway} from a {@link org.jclouds.aws.ec2.domain.VPC}
+    *
+    * @param region Region where the VPC exists
+    * @param internetGatewayId ID of the gateway to detach
+    * @param vpcId The ID of the VPC
+    * @param options Options for the request
+    */
+   @Named("DetachInternetGateway")
+   @POST
+   @FormParams(keys = ACTION, values = "DetachInternetGateway")
+   @XMLResponseParser(ReturnValueHandler.class)
+   @Fallback(FalseOnNotFoundOr404.class)
+   Boolean detachInternetGateway(
+      @EndpointParam(parser = RegionToEndpointOrProviderIfNull.class) @Nullable String region,
+      @FormParam("InternetGatewayId") String internetGatewayId,
+      @FormParam("VpcId") String vpcId,
+      InternetGatewayOptions... options);
```

Comment:
I know it is already a pattern in AWS, but it is legacy and wrong: I'd change the varargs parameter with an overloaded variant: one method with the options parameter and one without it. We actually don't support passing more than one options object, so the varargs parameter is not a good method signature. It's not the right way to model optional parameters.